src/common/artprov.cpp

```diff
@@ -165,42 +165,29 @@ class wxBitmapBundleImplArt : public wxBitmapBundleImpl
 
     virtual wxSize GetPreferredBitmapSizeAtScale(double scale) const wxOVERRIDE
     {
-        // We have no preferred sizes.
-        return m_sizeDefault*scale;
+        // Use the standard logic for integer-factor upscaling.
+        return DoGetPreferredSize(scale);
     }
 
     virtual wxBitmap GetBitmap(const wxSize& size) wxOVERRIDE
     {
         return wxArtProvider::GetBitmap(m_artId, m_artClient, size);
     }
 
-private:
-    static wxSize GetValidSize(const wxArtID& id,
-                               const wxArtClient& client,
-                               const wxSize& size)
+protected:
+    virtual double GetNextAvailableScale(size_t& i) const wxOVERRIDE
     {
-        // If valid size is provided, just use it.
-        if ( size != wxDefaultSize )
-            return size;
-
-        // Otherwise, try to get the size we'd use without creating a bitmap
-        // immediately.
-        const wxSize sizeHint = wxArtProvider::GetSizeHint(client);
-        if ( sizeHint != wxDefaultSize )
-            return sizeHint;
-
-        // If we really have to, do create a bitmap just to get its size. Note
-        // we need the size in logical pixels here, it will be scaled later if
-        // necessary, so use GetDIPSize() and not GetSize().
-        const wxBitmap bitmap = wxArtProvider::GetBitmap(id, client);
-        if ( bitmap.IsOk() )
-            return bitmap.GetDIPSize();
-
-        // We really need some default size, so just return this hardcoded
-        // value if all else fails -- what else can we do.
-        return wxSize(16, 16);
+        // Unfortunately we don't know what bitmap sizes are available here as
+        // there is simply nothing in wxArtProvider API that returns this (and
+        // adding something to the API doesn't make sense as all this is only
+        // used for compatibility with the existing custom art providers -- new
+        // ones should just override CreateBitmapBundle() directly), so we only
+        // return the original bitmap size, but hope that perhaps the provider
+        // will have a x2 version too, when our GetBitmap() is called.
+        return i++ ? 0.0 : 1.0;
```

Comment:
I missed this in my initial review, but apparently was bothered by it subconsciously enough to wake up knowing it was the cause of my scaling issues. This shouldn't read "1.0", but should use the bitmap's scaling factor.

Reply:
Oops, yes, of course, thanks for finding this!